File: trees2.py
# problem 1 : 106. Construct Binary Tree from Inorder and Postorder Traversal : https://leetcode.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal/
# Time Complexity : O(n)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        if not postorder:
            return None
        indict = {val:ind for ind,val in enumerate(inorder)}
        def makeTree(l, r):
            if l > r:
                return None
            val = postorder.pop()
            node= TreeNode(val)
            ind = indict[val]
            node.right = makeTree(ind+1, r)
            node.left = makeTree(l, ind-1)
            return node
        return makeTree(0, len(inorder)-1)



# problem 2 : 129. Sum Root to Leaf Numbers : https://leetcode.com/problems/sum-root-to-leaf-numbers/
# Time Complexity : O(n)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: TreeNode) -> int:
        ## iterative way
        total = 0
        stack = [(root,0)]
        while stack:
            root, curr = stack.pop()
            if root:
                curr = curr * 10 + root.val
                if not root.left and not root.right:
                    total += curr
                else:
                    stack.append((root.left, curr))
                    stack.append((root.right, curr))
        return total
        
        ## simple way with strings
#         if not root:
#             return 0
#         self.total = 0
        
#         def getSum(node,num):
#             #print(num)
#             if not node.left and not node.right:
#                 self.total += int(num+str(node.val))
#                 return
#             if node.left:
#                 getSum(node.left, num+str(node.val))
#             if node.right:
#                 getSum(node.right, num+str(node.val))
#         getSum(root, '')
#         return self.total
               

File: Trees2.java
//Time Complexity O(N) and Space Complexity O(N)
class Solution {
    int postIdx;
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if(inorder == null && postorder == null) return null;
        if(inorder.length != postorder.length) return null;
        postIdx=postorder.length-1;
        HashMap<Integer,Integer> map = new HashMap<>();
        for(int i=0;i<inorder.length;i++){
            map.put(inorder[i],i);
        }

        return helper(map,0,inorder.length-1,postorder);
        
    }

    public TreeNode helper(HashMap<Integer,Integer> map, int start,int end, int[] postorder){

        if(start>end) return null;

        int rootVal = postorder[postIdx];
        postIdx--;
        System.out.println(postIdx);
        TreeNode root = new TreeNode(rootVal);

        int rootIdx = map.get(rootVal);

        root.right = helper(map,rootIdx+1,end,postorder);
        root.left = helper(map,start,rootIdx-1,postorder);
        

        return root;

    }
}




//Sum-to root Time Complexity O(N) and space complexity O(H)
class Solution {
    int sum;
    public int sumNumbers(TreeNode root) {

        return helper(root,0,0);
        
    }


    public int helper(TreeNode root, int curr, int sum){
        if(root==null) return 0;

        curr = curr*10+root.val;
        int left = helper(root.left,curr,sum);
        int right = helper(root.right,curr,sum);
        if(root.left==null && root.right==null){
            sum+=curr;
            return sum;
        }
        
        

        return left+right;
    }
}
